cmd/registry-replacer/main.go

```diff
@@ -0,0 +1,210 @@
+	"net/http"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/config"
+)
+
+func main() {
+	configDir := flag.String("config-dir", "", "The directory with the ci-operator configs")
+	flag.Parse()
+	if *configDir == "" {
+		logrus.Fatal("--config-dir arg is required")
+	}
+
+	var errs []error
+	errLock := &sync.Mutex{}
+	wg := sync.WaitGroup{}
+	if err := config.OperateOnCIOperatorConfigDir(
+		*configDir,
+		func(config *api.ReleaseBuildConfiguration, info *config.Info) error {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				if err := replacer(githubFileGetterFactory, ioutil.WriteFile)(config, info); err != nil {
+					errLock.Lock()
+					errs = append(errs, err)
+					errLock.Unlock()
+				}
+			}()
+			return nil
+		},
+	); err != nil {
+		logrus.WithError(err).Fatal("Failed to operate on ci-operator-config")
+	}
+	wg.Wait()
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		logrus.WithError(err).Fatal("Encountered errors")
+	}
+}
+
+// replacer ensures replace directives are in place. It fetches the files via http because using git
+// en masse easily kills a developer laptop whereas the http calls are cheap and can be parallelized without
+// bounds.
+func replacer(
+	githubFileGetterFactory func(org, repo, branch string) githubFileGetter,
+	writer func(name string, data []byte, mode os.FileMode) error,
+) func(*api.ReleaseBuildConfiguration, *config.Info) error {
+	return func(config *api.ReleaseBuildConfiguration, info *config.Info) error {
+		if !referencesDockerfile(config.Images) {
+			return nil
+		}
+
+		originalConfig, err := yaml.Marshal(config)
+		if err != nil {
+			return fmt.Errorf("failed to marshal config for comparison: %w", err)
+		}
+
+		for idx := range config.Images {
+			foundTags, err := ensureReplacement(&config.Images[idx], githubFileGetterFactory(info.Org, info.Repo, info.Branch))
+			if err != nil {
+				return fmt.Errorf("failed to ensure replacements: %w", err)
+			}
+			for _, foundTag := range foundTags {
+				if config.BaseImages == nil {
+					config.BaseImages = map[string]api.ImageStreamTagReference{}
+				}
+				config.BaseImages[foundTag.String()] = api.ImageStreamTagReference{
+					Namespace: foundTag.org,
+					Name:      foundTag.repo,
+					Tag:       foundTag.tag,
+				}
+			}
+		}
+
+		newConfig, err := yaml.Marshal(config)
+		if err != nil {
+			return fmt.Errorf("failed to marshal new config: %w", err)
+		}
+
+		// Avoid filesystem access if possible
+		if bytes.Equal(originalConfig, newConfig) {
+			return nil
+		}
+
+		if err := writer(info.Filename, newConfig, 0644); err != nil {
+			return fmt.Errorf("faild to write %s: %w", info.Filename, err)
+		}
+
+		return nil
+	}
+}
+
+func referencesDockerfile(images []api.ProjectDirectoryImageBuildStepConfiguration) bool {
```

Comment:
`context_dir` also should be handled here, as that references `context_dir/Dockerfile`

Reply:
Done